src/java_bytecode/java_object_factory.cpp

```diff
@@ -110,42 +257,155 @@ void gen_nondet_init(
 
       if(name=="@class_identifier")
       {
+	if(skip_classid)
+	  continue;
+
         irep_idt qualified_clsid="java::"+as_string(class_identifier);
-        constant_exprt ci(qualified_clsid, string_typet());
+	constant_exprt ci(qualified_clsid,string_typet());
 
-        code_assignt code(me, ci);
-        init_code.copy_to_operands(code);
+	code_assignt code(me, ci);
+        code.add_source_location() = loc;
+	init_code.copy_to_operands(code);
       }
       else if(name=="@lock")
       {
         code_assignt code(me, gen_zero(me.type()));
+        code.add_source_location() = loc;
         init_code.copy_to_operands(code);
       }
       else
       {
         assert(!name.empty());
 
         bool _is_sub = name[0]=='@';
+#if 0
         irep_idt _class_identifier=
           _is_sub?(class_identifier.empty()?struct_tag:class_identifier):"";
+#endif
 
-        gen_nondet_init(
-          me, init_code, ns, recursion_set, _is_sub, _class_identifier);
+        gen_nondet_init(me, _is_sub, class_identifier, loc, false, create_dynamic_objects);
       }
     }
-
     recursion_set.erase(struct_tag);
   }
   else
   {
     side_effect_expr_nondett se=side_effect_expr_nondett(type);
 
     code_assignt code(expr, se);
+    code.add_source_location() = loc;
     init_code.copy_to_operands(code);
   }
 }
+
+// Borrowed from java_bytecode_convert.cpp -- todo find a sensible place to factor this.
+static constant_exprt as_number(const mp_integer value, const typet &type)
+{
+  const size_t java_int_width(type.get_unsigned_int(ID_width));
+  const std::string significant_bits(integer2string(value, 2));
+  std::string binary_width(java_int_width - significant_bits.length(), '0');
+  return constant_exprt(binary_width += significant_bits, type);
+}
+
+void gen_nondet_state::gen_nondet_array_init(const exprt &expr, const source_locationt &loc)
```

Comment:
This code is a bit obscure. Please explain in the function comment block the code that you are generating.